agent/xds/xdscommon/xdscommon.go

```diff
@@ -97,70 +127,91 @@ func MakePluginConfiguration(cfgSnap *proxycfg.ConfigSnapshot) PluginConfigurati
 
 	switch cfgSnap.Kind {
 	case structs.ServiceKindConnectProxy:
-		connectProxies := make(map[proxycfg.UpstreamID]struct{})
+		kind = api.ServiceKindConnectProxy
+		outgoingKindByService := make(map[api.CompoundServiceName]api.ServiceKind)
 		for uid, upstreamData := range cfgSnap.ConnectProxy.WatchedUpstreamEndpoints {
+			sn := upstreamIDToCompoundServiceName(uid)
+
 			for _, serviceNodes := range upstreamData {
-				// Lambdas and likely other integrations won't be attached to nodes.
-				// After agentless, we may need to reconsider this.
-				if len(serviceNodes) == 0 {
-					connectProxies[uid] = struct{}{}
-				}
 				for _, serviceNode := range serviceNodes {
-					if serviceNode.Service.Kind == structs.ServiceKindTypical || serviceNode.Service.Kind == structs.ServiceKindConnectProxy {
-						connectProxies[uid] = struct{}{}
+					if serviceNode.Service == nil {
+						continue
 					}
+					// Store the upstream's kind, and for ServiceKindTypical we don't do anything because we'll default
+					// any unset upstreams to ServiceKindConnectProxy below.
+					switch serviceNode.Service.Kind {
+					case structs.ServiceKindTypical:
+					default:
+						outgoingKindByService[sn] = api.ServiceKind(serviceNode.Service.Kind)
+					}
+					// We only need the kind from one instance, so break once we find it.
+					break
 				}
 			}
 		}
 
 		// TODO(peering): consider PeerUpstreamEndpoints in addition to DiscoveryChain
-
+		// These are the discovery chains for upstreams which have the Envoy Extensions applied to the local service.
 		for uid, dc := range cfgSnap.ConnectProxy.DiscoveryChain {
-			if _, ok := connectProxies[uid]; !ok {
-				continue
-			}
-
-			serviceConfigs[upstreamIDToCompoundServiceName(uid)] = ServiceConfig{
-				Kind:            api.ServiceKindConnectProxy,
-				EnvoyExtensions: convertEnvoyExtensions(dc.EnvoyExtensions),
-			}
-
 			compoundServiceName := upstreamIDToCompoundServiceName(uid)
+			extensionConfigs[compoundServiceName] = convertEnvoyExtensions(dc.EnvoyExtensions)
```

Comment:
We eventually might want to convert this to a `ToAPI` function. This is a [pretty common patter](https://github.com/hashicorp/consul/blob/1f89de051cbe4f1b8cc827d013feb44c13c7cbb6/agent/structs/connect_proxy_config.go#L95-L97).

Reply:
I've added a ToAPI function for structs.EnvoyExtensions which is called by convertEnvoyExtensions. I didn't find a good way to convert dc.EnvoyExtensions from type []structs.EnvoyExtension --> structs.EnvoyExtensions, where structs.EnvoyExtensions is defined to be []structs.EnvoyExtension